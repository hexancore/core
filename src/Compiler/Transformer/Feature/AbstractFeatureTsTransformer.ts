import ts from 'typescript';
import type { FeatureMap, FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";
import { ModuleClassTsTransformer } from '../ModuleClassTsTransformer';
import type { TsImportHelper } from '../TsImportHelper';
import type { FeatureTransformContext } from './FeatureTransformContext';
import type { FeatureSourcePath } from "../../../Util/Feature/FeatureModuleDiscoverer";

export interface FeatureTsTransformerHelpers {
  importHelper: TsImportHelper;
  moduleClassTransformer: ModuleClassTsTransformer;
  features: FeatureMap;
}

export abstract class AbstractFeatureTsTransformer {
  protected importHelper: TsImportHelper;
  protected moduleClassTransformer: ModuleClassTsTransformer;
  protected features: FeatureMap;

  public constructor(helpers: FeatureTsTransformerHelpers) {
    this.importHelper = helpers.importHelper;
    this.moduleClassTransformer = helpers.moduleClassTransformer;
    this.features = helpers.features;
  }

  public abstract transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile;

  public abstract supports(featureSourcePath: FeatureSourcePath, feature: FeatureMeta): boolean;

}