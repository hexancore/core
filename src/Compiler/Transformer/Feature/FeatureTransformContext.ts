import type ts from "typescript";
import type { FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";

export interface FeatureTransformContext {
  feature: FeatureMeta;
  sourceFilePathWithoutRoot: string;
  tsContext: ts.TransformationContext;
  diagnostics: ts.Diagnostic[];
}