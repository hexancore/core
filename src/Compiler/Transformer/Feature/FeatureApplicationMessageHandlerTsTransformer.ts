import { FeatureApplicationCommandMeta, type FeatureApplicationQueryMeta } from '../../../Util/Feature/Meta';
import ts from "typescript";
import { type FeatureMeta } from "../../../Util/Feature/Meta";
import { AbstractFeatureTsTransformer } from "./AbstractFeatureTsTransformer";
import type { FeatureTransformContext } from "./FeatureTransformContext";
import type { FeatureSourcePath } from "../../../Util/Feature/FeatureModuleDiscoverer";
import { ImportDeclarationWrapper } from "../Helper/ImportDeclarationWrapper";

interface VisitContext extends FeatureTransformContext {
  meta: FeatureApplicationCommandMeta | FeatureApplicationQueryMeta;
  handlerDecoratorName: string;
  cqrsImport: ImportDeclarationWrapper;
}

export class FeatureApplicationMessageHandlerTsTransformer extends AbstractFeatureTsTransformer {

  public supports(featureSourcePath: FeatureSourcePath, feature: FeatureMeta): boolean {
    const meta = feature.applicationCqrsHandlerMap.get(featureSourcePath.localSourcePath);
    return !!meta;
  }

  public transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile {
    const messageMeta = context.feature.applicationCqrsHandlerMap.get(context.featureSourcePath.localSourcePath)!;
    const handlerDecoratorName = messageMeta instanceof FeatureApplicationCommandMeta ? "CommandHandler" : "QueryHandler";
    const cqrsImport = ImportDeclarationWrapper.create(handlerDecoratorName, "@nestjs/cqrs", context.tsContext);
    const visitContext: VisitContext = {
      ...context,
      meta: messageMeta,
      handlerDecoratorName: handlerDecoratorName,
      cqrsImport
    };

    const transformed = ts.factory.updateSourceFile(source, [
      cqrsImport.declaration,
      ...ts.visitNodes(
        source.statements,
        this.visitSourceStatement.bind(this, visitContext),
        (node): node is ts.Statement => ts.isStatement(node)
      )
    ]);

    return transformed;
  }

  private visitSourceStatement(visitContext: VisitContext, node: ts.Statement) {
    if (ts.isClassDeclaration(node)) {
      const handlerDecorator = ts.factory.createDecorator(ts.factory.createCallExpression(
        visitContext.cqrsImport.get(visitContext.handlerDecoratorName),
        undefined,
        [ts.factory.createIdentifier(visitContext.meta.className)]
      ));
      return ts.factory.updateClassDeclaration(
        node,
        [handlerDecorator, ...node.modifiers ?? []],
        node.name,
        node.typeParameters,
        node.heritageClauses,
        node.members
      );
    }

    return node;
  }
}