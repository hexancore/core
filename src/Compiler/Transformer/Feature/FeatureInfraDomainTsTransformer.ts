import ts from 'typescript';
import type { FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";
import { AbstractFeatureTsTransformer } from './AbstractFeatureTsTransformer';
import { TsTransfromerHelper } from '../TsTransformerHelper';
import type { AddImportTransformDef } from '../ModuleClassTsTransformer';
import type { FeatureTransformContext } from './FeatureTransformContext';

export class FeatureInfraDomainModuleTsTransformer extends AbstractFeatureTsTransformer {

  public transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile {
    const feature = context.feature;
    const domainErrorsClassName = feature.name + "DomainErrors";

    const imports: AddImportTransformDef[] = [
      { name: "DomainInfraModuleHelper", importModuleSpecifier: '@hexancore/core' },
      { name: domainErrorsClassName, importModuleSpecifier: '../../Domain' }
    ];

    const repos: string[] = [];

    for (const r of feature.domain.aggregateRoots) {
      repos.push(r.infraRepositoryName);
      imports.push(
        { name: r.name, importModuleSpecifier: `../../Domain/${r.name}/${r.name}` },
        { name: r.infraRepositoryName, importModuleSpecifier: `./${r.name}/${r.infraRepositoryName}` }
      );

      for (const e of r.entities) {
        repos.push(e.infraRepositoryName);
        imports.push({ name: e.infraRepositoryName, importModuleSpecifier: `./${r.name}/${e.infraRepositoryName}` });
      }
    }

    return this.moduleClassTransformer.transform({
      imports,
      extraStatementProvider(importedIdentifierMapper) {
        const classIdentifier = importedIdentifierMapper("DomainInfraModuleHelper");
        const methodIdentifier = ts.factory.createIdentifier("createMeta");

        const optionsObject = ts.factory.createObjectLiteralExpression([
          ts.factory.createPropertyAssignment("featureName", ts.factory.createStringLiteral(feature.name)),
          ts.factory.createPropertyAssignment("aggregateRootCtrs", ts.factory.createArrayLiteralExpression(feature.domain.aggregateRoots.map((r) => ts.factory.createIdentifier(r.name)))),
          ts.factory.createPropertyAssignment("domainErrors", ts.factory.createIdentifier(domainErrorsClassName)),
        ]);

        const createMeta = TsTransfromerHelper.createConstStatement("HcDomainInfraModuleMetaExtra", ts.factory.createCallExpression(
          ts.factory.createPropertyAccessExpression(classIdentifier, methodIdentifier),
          undefined,
          [optionsObject]
        ));

        return [
          TsTransfromerHelper.createConstStatement("HcDomainInfraAggrgateRootRepositories",
            ts.factory.createArrayLiteralExpression(repos.map((r) => ts.factory.createIdentifier(r)))
          ),
          createMeta
        ];
      },

      extraMetaProvider() {
        return ts.factory.createIdentifier("HcDomainInfraModuleMetaExtra");
      },
      source,
      context: context.tsContext,
    });
  }

  public supports(sourcefilePath: string, feature: FeatureMeta): boolean {
    return sourcefilePath.endsWith("DomainInfraModule.ts");
  }
}