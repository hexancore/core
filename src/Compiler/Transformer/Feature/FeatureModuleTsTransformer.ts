import ts from 'typescript';
import type {FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";
import { AbstractFeatureTsTransformer } from "./AbstractFeatureTsTransformer";
import type { ProviderModuleMetaTransformDef } from '../ModuleClassTsTransformer';
import type { FeatureTransformContext } from './FeatureTransformContext';
import type { FeatureApplicationMessageMeta } from '../../../Util/Feature/Meta';
import type { FeatureSourcePath } from "../../../Util/Feature/FeatureModuleDiscoverer";

/**
 * Adding automatic injection of message handlers, services, infra module to `[Feature]Module` source.
 * Less write, more fun !
 */
export class FeatureModuleTsTransformer extends AbstractFeatureTsTransformer {

  public supports(featureSourcePath: FeatureSourcePath, feature: FeatureMeta): boolean {
    return featureSourcePath.localSourcePath.endsWith(feature.name + "Module.ts");
  }

  public transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile {
    const feature = context.feature;

    const messageHandlersProviders: ProviderModuleMetaTransformDef[] = [];
    messageHandlersProviders.push(...this.createMessageHandlerProviders(feature.application.commands));
    messageHandlersProviders.push(...this.createMessageHandlerProviders(feature.application.queries));

    return this.moduleClassTransformer.transform({
      imports: [],
      meta: {
        imports: [],
        providers: [...messageHandlersProviders, ...this.createServiceProviders(feature)],
      },
      source,
      context: context.tsContext,
    });
  }

  private createMessageHandlerProviders(messages: FeatureApplicationMessageMeta[]): ProviderModuleMetaTransformDef[] {
    const providers: ProviderModuleMetaTransformDef[] = [];
    for (const m of messages) {
      const importPath = `./${m.path}/${m.handlerClass}`;
      providers.push({
        addToExports: false,
        name: m.handlerClass,
        importFrom: importPath
      });
    }

    return providers;
  }

  private createServiceProviders(feature: FeatureMeta): ProviderModuleMetaTransformDef[] {
    const providers: ProviderModuleMetaTransformDef[] = [];
    for (const s of feature.application.services) {
      if (!s.isInjectable) {
        continue;
      }

      providers.push({
        addToExports: false,
        name: s.name,
        importFrom: `./${s.path}`
      });
    }

    return providers;
  }
}