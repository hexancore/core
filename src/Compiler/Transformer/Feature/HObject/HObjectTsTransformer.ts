import { HObjectPropertyExtractor } from './HObjectPropertyExtractor';
import ts from "typescript";
import type { FeatureMeta } from "../../../../Util/Feature/Meta/FeatureMeta";
import { AbstractFeatureTsTransformer, type FeatureTsTransformerHelpers } from "../AbstractFeatureTsTransformer";
import type { FeatureTransformContext } from "../FeatureTransformContext";
import { HObjectParseTsFactory } from "./HObjectParseTsFactory";
import { HObjectToJSONTsFactory } from "./HObjectToJSONTsFactory";
import type { FeatureSourcePath } from "../../../../Util/Feature/FeatureModuleDiscoverer";
import { HObjectToConstructorTsFactory } from "./HObjectConstructorTsFactory";
import { ImportDeclarationWrapper } from "../../Helper/ImportDeclarationWrapper";
import { HObjectKind } from "@/Util/Feature/Meta";

interface VisitContext extends FeatureTransformContext {
  source: ts.SourceFile;
  importNameToDeclarationMap: Map<string, ImportDeclarationWrapper>;
  hCommonImportDecl?: ImportDeclarationWrapper;
}

const HObjectHCommonImports = [
  'R',
  'OK',
  'type JsonObjectType',
  'type PlainParsableHObjectType',
  'type PlainParseError',
  'PlainParseHelper',
  'InvalidTypePlainParseIssue',
  'HObjectTypeMeta',
  'PlainParseIssue',
  'TooBigPlainParseIssue',
  'TooSmallPlainParseIssue'
];

export class HObjectTsTransformer extends AbstractFeatureTsTransformer {

  private propertyExtractor: HObjectPropertyExtractor;
  private constructorTsFactory: HObjectToConstructorTsFactory;
  private parseTsFactory: HObjectParseTsFactory;
  private toJSONTsFactory: HObjectToJSONTsFactory;

  public constructor(helpers: FeatureTsTransformerHelpers) {
    super(helpers);

    this.propertyExtractor = new HObjectPropertyExtractor();
    this.constructorTsFactory = new HObjectToConstructorTsFactory(),
      this.parseTsFactory = new HObjectParseTsFactory();
    this.toJSONTsFactory = new HObjectToJSONTsFactory();
  }

  public supports(featureSourcePath: FeatureSourcePath, feature: FeatureMeta): boolean {
    const meta = feature.hObjectMap.get(featureSourcePath.localSourcePath);
    if (meta && meta.kind !== HObjectKind.AggregateRoot && meta.kind !== HObjectKind.Entity) {
      return true;
    }

    return false;
  }

  public transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile {
    const visitContext: VisitContext = {
      ...context,
      importNameToDeclarationMap: new Map(),
      source,
    };

    const transformed = ts.factory.updateSourceFile(source, ts.visitNodes(
      source.statements,
      this.visitSourceStatment.bind(this, visitContext),
      (node): node is ts.Statement => ts.isStatement(node))
    );

    return transformed;
  }

  private visitSourceStatment(visitContext: VisitContext, node: ts.Statement) {
    if (ts.isImportDeclaration(node)) {
      return this.visitImportDeclaration(visitContext, node);
    }

    if (ts.isClassDeclaration(node)) {
      if (!visitContext.hCommonImportDecl) {
        visitContext.hCommonImportDecl = ImportDeclarationWrapper.create(HObjectHCommonImports, '@hexancore/common', visitContext.tsContext);
        return [
          visitContext.hCommonImportDecl.declaration,
          this.updateClassDeclaration(node, visitContext as any)
        ];
      }

      return this.updateClassDeclaration(node, visitContext as any);
    }

    return node;
  }

  private visitImportDeclaration(visitContext: VisitContext, node: ts.ImportDeclaration) {
    const wrapper = ImportDeclarationWrapper.from(node, visitContext.tsContext);

    if (this.isHexancoreCommonImport(visitContext, node)) {
      visitContext.hCommonImportDecl = wrapper;
      return this.extendHexancoreCommonImportDecl(visitContext.hCommonImportDecl);
    }

    wrapper.importNames.forEach(name => {
      visitContext.importNameToDeclarationMap.set(name, wrapper);
    });

    return wrapper.declaration;
  }

  private isHexancoreCommonImport(visitContext: VisitContext, node: ts.ImportDeclaration): boolean {
    const moduleSpecifier = node.moduleSpecifier;
    return !visitContext.hCommonImportDecl && ts.isStringLiteral(moduleSpecifier) && moduleSpecifier.text === '@hexancore/common';
  }

  private extendHexancoreCommonImportDecl(decl: ImportDeclarationWrapper): ts.ImportDeclaration {
    const newImports = HObjectHCommonImports.filter(i => !decl.hasNamedAccess(i));
    return decl.addNamedImports(newImports);
  }

  private updateClassDeclaration(node: ts.ClassDeclaration, visitContext: Required<VisitContext>) {
    const properties = this.propertyExtractor.extract(node, visitContext.source, visitContext.diagnostics);
    const hObjectClassName = node.name!.text;

    const newMembers = [
      this.createHObjectMetaClassProperty(hObjectClassName, visitContext),
      ...node.members,
    ];

    if (properties.length > 0) {
      newMembers.push(this.constructorTsFactory.create(properties));
      newMembers.push(this.parseTsFactory.create(hObjectClassName, properties, visitContext.hCommonImportDecl));
      newMembers.push(this.toJSONTsFactory.create(hObjectClassName, properties, visitContext.hCommonImportDecl));
    }

    return ts.factory.updateClassDeclaration(
      node,
      node.modifiers,
      node.name,
      node.typeParameters,
      node.heritageClauses,
      ts.factory.createNodeArray(newMembers)
    );
  }

  private createHObjectMetaClassProperty(hObjClassName: string, visitContext: VisitContext) {
    const featureHObjMeta = visitContext.feature.hObjectMap.get(visitContext.featureSourcePath.localSourcePath)!;

    const hObjTypeMetaCreateParameters = [
      ts.factory.createStringLiteral(visitContext.feature.name),
      ts.factory.createStringLiteral(featureHObjMeta.context),
      ts.factory.createStringLiteral(featureHObjMeta.kind + ''),
      ts.factory.createStringLiteral(featureHObjMeta.name),
      ts.factory.createIdentifier(hObjClassName)
    ];

    const hobjectTypeMetaCreateExp = ts.factory.createCallExpression(
      ts.factory.createPropertyAccessExpression(
        ts.factory.createIdentifier("HObjectTypeMeta"),
        ts.factory.createIdentifier(featureHObjMeta.layer)
      ),
      undefined,
      hObjTypeMetaCreateParameters
    );

    return ts.factory.createPropertyDeclaration(
      [ts.factory.createModifier(ts.SyntaxKind.PublicKeyword), ts.factory.createModifier(ts.SyntaxKind.StaticKeyword)],
      "HOBJ_META",
      undefined,
      undefined,
      hobjectTypeMetaCreateExp
    );
  }
}