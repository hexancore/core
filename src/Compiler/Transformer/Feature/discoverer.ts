import { FeatureModuleDiscoverer } from "../../../Util/Feature/FeatureModuleDiscoverer";
// node ./lib/Compiler/Transformer/Feature/discoverer.js ./test/helper/libs/test-lib/src
async function run() {
  if (process.argv.length < 3) {
    throw new Error("Wrong execute, not given sourceRoot argument");
  }
  const sourceRoot = process.argv[2];
  const discoverer = new FeatureModuleDiscoverer(sourceRoot);
  const features = await discoverer.discoverAll();
  console.log(JSON.stringify(Array.from(features.entries())));
}

run();