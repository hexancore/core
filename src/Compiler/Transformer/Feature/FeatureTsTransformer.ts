import { FsHelper } from "@/Util/Filesystem/FsHelper";
import { execFileSync } from "node:child_process";
import ts from 'typescript';
import { FeatureMap, FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";
import { TsTransfromerHelper } from "../TsTransformerHelper";
import type { AbstractFeatureTsTransformer } from "./AbstractFeatureTsTransformer";
import { FeatureInfraDomainModuleTsTransformer } from "./FeatureInfraDomainTsTransformer";
import { FeatureModuleTsTransformer } from "./FeatureModuleTsTransformer";
import path from "node:path";
import { FeatureModuleDiscoverer, type FeatureSourcePath } from "../../../Util/Feature/FeatureModuleDiscoverer";
import { TsImportHelper } from "../TsImportHelper";
import { ModuleClassTsTransformer } from "../ModuleClassTsTransformer";
import type { FeatureTransformContext } from "./FeatureTransformContext";
import { HObjectTsTransformer } from "./HObject/HObjectTsTransformer";


/**
 * Adding automatic injection of message handlers, services, infra module to `[Feature]Module` source.
 * Less write, more fun !
 */
export class FeatureTsTransformer {
  private transformers: AbstractFeatureTsTransformer[];

  public constructor(
    private sourceRoot: string,
    private features: FeatureMap,
    importHelper: TsImportHelper,
  ) {

    const helpers = {
      importHelper: importHelper,
      moduleClassTransformer: new ModuleClassTsTransformer(importHelper),
      features: this.features,
    };

    this.transformers = [
      new FeatureInfraDomainModuleTsTransformer(helpers),
      new FeatureModuleTsTransformer(helpers),
      new HObjectTsTransformer(helpers),
    ];
  }

  public static create(sourceRoot: string, compilerDir?: string, features?: FeatureMap, needFixImportAccess = true): FeatureTsTransformer {
    sourceRoot = FsHelper.normalizePathSep(sourceRoot);
    compilerDir = compilerDir ?? FsHelper.normalizePathSep(path.dirname(path.dirname(__dirname)));
    features = features ?? this.loadFeaturesMapSync(sourceRoot, compilerDir);

    const importFromMapper = TsTransfromerHelper.createImportFromMapper(sourceRoot);
    const importHelper = new TsImportHelper(importFromMapper, needFixImportAccess);

    return new this(sourceRoot, features, importHelper);
  }

  private static loadFeaturesMapSync(sourceRoot: string, compilerDir: string): Map<string, FeatureMeta> {
    // TS transformer can't use async, but we need async code and execSync allows run async code :) Hack of day !
    const output = execFileSync('node', [`${compilerDir}/Transformer/Feature/discoverer.js`, sourceRoot], {
      windowsHide: true,
      shell: false,
      timeout: 60 * 1000
    });
    const featuresPlain = JSON.parse(output.toString());
    return FeatureMeta.parseMap(featuresPlain);
  }

  public transform(source: ts.SourceFile, context: ts.TransformationContext): ts.SourceFile {
    const featureSourceTransformContext = this.createFeatureSourceTransformContext(source, context);
    if (!featureSourceTransformContext) {
      return source;
    }


    for (const t of this.transformers) {
      if (t.supports(featureSourceTransformContext.featureSourcePath, featureSourceTransformContext.feature)) {
        const transformed = t.transform(source, featureSourceTransformContext);
        return transformed;
      }
    }

    return source;
  }

  private createFeatureSourceTransformContext(source: ts.SourceFile, context: ts.TransformationContext): FeatureTransformContext | null {
    const sourceFilePath = FsHelper.normalizePathSep(source.fileName);
    const featureSourcePath = FeatureModuleDiscoverer.extractFeatureNameFromPath(this.sourceRoot, sourceFilePath);
    if (!featureSourcePath) {
      return null;
    }

    const feature = this.features.get(featureSourcePath.featureName)!;
    return {
      feature,
      featureSourcePath,
      tsContext: context,
      diagnostics: [],
    };
  }

  public supports(sourcePath: FeatureSourcePath): boolean {
    const feature = this.features.get(sourcePath.featureName)!;
    for (const t of this.transformers) {
      if (t.supports(sourcePath, feature)) {
        return true;
      }
    }

    return false;
  }
}