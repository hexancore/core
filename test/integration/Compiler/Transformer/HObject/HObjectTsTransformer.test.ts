/**
 * @group integration
 */

import { TsTransformerTestHelper } from "@/Compiler/Test/TsTransformerTestHelper";
import { FeatureTsTransformer } from "@/Compiler/Transformer/Feature/FeatureTsTransformer";
import { HObjectTsTransformer } from "@/Compiler/Transformer/Feature/HObject/HObjectTsTransformer";
import { LIBS_DIRNAME } from "@test/libs";
import type { TransformerFactory, SourceFile } from "typescript";

describe(HObjectTsTransformer.constructor.name, () => {
  const helper = TsTransformerTestHelper.createFromTsConfig(process.cwd() + "/tsconfig.json");
  const sourceRoot = LIBS_DIRNAME + "/test-lib/src";
  let transformer: FeatureTsTransformer;

  let transformerFactory: TransformerFactory<SourceFile>;

  beforeAll(() => {

    const compilerRoot = process.cwd() + "/lib/Compiler";
    transformer = FeatureTsTransformer.create(sourceRoot, compilerRoot);
    transformerFactory = helper.createTransformerFactory(transformer.transform.bind(transformer));
  });

  test("transform: Command", () => {
    const sourceFilePath = `${sourceRoot}/Book/Application/Book/Command/Create/BookCreateCommand.ts`;

    const out = helper.transformExistingAndReturnAsString(sourceFilePath, [transformerFactory]);

    expect(out).toMatchSnapshot();
  });

  test("transform: Query", () => {
    const sourceFilePath = `${sourceRoot}/Book/Application/Book/Query/GetById/BookGetByIdQuery.ts`;

    const out = helper.transformExistingAndReturnAsString(sourceFilePath, [transformerFactory]);

    expect(out).toMatchSnapshot();
  });

  test("transform: DTO", () => {
    const sourceFilePath = `${sourceRoot}/Book/Application/Book/Dto/TestTransformDto.ts`;

    const out = helper.transformExistingAndReturnAsString(sourceFilePath, [transformerFactory]);

    expect(out).toMatchSnapshot();
  });

  test("transform: ValueObject", () => {
    const sourceFilePath = `${sourceRoot}/Book/Domain/Book/Shared/ValueObject/BookId.ts`;

    const out = helper.transformExistingAndReturnAsString(sourceFilePath, [transformerFactory]);

    expect(out).toMatchSnapshot();
  });

});


