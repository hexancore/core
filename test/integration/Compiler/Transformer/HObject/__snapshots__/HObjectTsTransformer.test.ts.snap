// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Function transform: Command 1`] = `
"import { HCommand, R, OK, type JsonObjectType, type PlainParsableHObjectType, type PlainParseError, PlainParseHelper, InvalidTypePlainParseIssue, HObjectTypeMeta, PlainParseIssue, TooBigPlainParseIssue, TooSmallPlainParseIssue } from "@hexancore/common";
export class BookCreateCommand extends HCommand<BookCreateCommand, void> {
    public static HOBJ_META = HObjectTypeMeta.application("Book", "Book", "Command", "Create", BookCreateCommand);
    public title!: string;
    public constructor(title: any) {
        this.title = title;
    }
    public static parse(plain: unknown): common_1.R<BookCreateCommand> {
        if (typeof plain !== "object") {
            return PlainParseHelper.HObjectIsNotObjectParseErr(BookCreateCommand as any, plain);
        }
        const p = plain as Record<keyof BookCreateCommand, unknown>;
        const issues: PlainParseIssue[] = [];
        const title = PlainParseHelper.parseString(p.title, "title", issues);
    }
    public static toJSON(): common_1.JsonObjectType<BookCreateCommand> {
        return {
            title: this.title
        };
    }
}
"
`;

exports[`Function transform: DTO 1`] = `
"import { Dto, v, RefId, R, OK, type JsonObjectType, type PlainParsableHObjectType, type PlainParseError, PlainParseHelper, InvalidTypePlainParseIssue, HObjectTypeMeta, PlainParseIssue, TooBigPlainParseIssue, TooSmallPlainParseIssue } from '@hexancore/common';
export class TestTransformDto extends Dto {
    public static HOBJ_META = HObjectTypeMeta.application("Book", "Book", "Dto", "TestTransformDto", TestTransformDto);
    public optionalField?: string;
    public numberField!: number;
    public stringField!: string;
    public booleanField!: boolean;
    public bigintField!: bigint;
    public primitiveArrayField!: string[];
    public uintField!: v.uint;
    public ruleWithArgsField!: v.int.between<-10, 100>;
    public ruleArrayField!: v.int.between<-10, 100>[];
    public ruleArrayWithItemsField!: v.int.between<-10, 100>[] & v.items.between<2, 5>;
    public hObjField!: RefId;
    public hObjArrayField!: RefId[];
    public constructor(numberField: any, stringField: any, booleanField: any, bigintField: any, primitiveArrayField: any, uintField: any, ruleWithArgsField: any, ruleArrayField: any, ruleArrayWithItemsField: any, hObjField: any, hObjArrayField: any, optionalField?: any) {
        this.numberField = numberField;
        this.stringField = stringField;
        this.booleanField = booleanField;
        this.bigintField = bigintField;
        this.primitiveArrayField = primitiveArrayField;
        this.uintField = uintField;
        this.ruleWithArgsField = ruleWithArgsField;
        this.ruleArrayField = ruleArrayField;
        this.ruleArrayWithItemsField = ruleArrayWithItemsField;
        this.hObjField = hObjField;
        this.hObjArrayField = hObjArrayField;
        this.optionalField = optionalField;
    }
    public static parse(plain: unknown): common_1.R<TestTransformDto> {
        if (typeof plain !== "object") {
            return PlainParseHelper.HObjectIsNotObjectParseErr(TestTransformDto as any, plain);
        }
        const p = plain as Record<keyof TestTransformDto, unknown>;
        const issues: PlainParseIssue[] = [];
        let optionalField;
        if (p.optionalField !== undefined) {
            optionalField = PlainParseHelper.parseString(p.optionalField, "optionalField", issues);
        }
        const stringField = PlainParseHelper.parseString(p.stringField, "stringField", issues);
        const primitiveArrayField = PlainParseHelper.parsePrimitiveArray(p.primitiveArrayField, PlainParseHelper.parseString, "primitiveArrayField", issues);
    }
    public static toJSON(): common_1.JsonObjectType<TestTransformDto> {
        return {
            optionalField: this.optionalField,
            numberField: this.numberField,
            stringField: this.stringField,
            booleanField: this.booleanField,
            bigintField: this.bigintField.toString(),
            primitiveArrayField: this.primitiveArrayField,
            uintField: this.uintField,
            ruleWithArgsField: this.ruleWithArgsField,
            ruleArrayField: this.ruleArrayField,
            ruleArrayWithItemsField: this.ruleArrayWithItemsField,
            hObjField: this.hObjField,
            hObjArrayField: this.hObjArrayField.map(item => item.toJSON())
        };
    }
}
"
`;

exports[`Function transform: Query 1`] = `
"import { HQuery, R, OK, type JsonObjectType, type PlainParsableHObjectType, type PlainParseError, PlainParseHelper, InvalidTypePlainParseIssue, HObjectTypeMeta, PlainParseIssue, TooBigPlainParseIssue, TooSmallPlainParseIssue } from "@hexancore/common";
import type { BookDto } from "../../Dto/BookDto";
export class BookGetByIdQuery extends HQuery<BookGetByIdQuery, BookDto> {
    public static HOBJ_META = HObjectTypeMeta.application("Book", "Book", "Query", "GetById", BookGetByIdQuery);
    public title!: string;
    public constructor(title: any) {
        this.title = title;
    }
    public static parse(plain: unknown): common_1.R<BookGetByIdQuery> {
        if (typeof plain !== "object") {
            return PlainParseHelper.HObjectIsNotObjectParseErr(BookGetByIdQuery as any, plain);
        }
        const p = plain as Record<keyof BookGetByIdQuery, unknown>;
        const issues: PlainParseIssue[] = [];
        const title = PlainParseHelper.parseString(p.title, "title", issues);
    }
    public static toJSON(): common_1.JsonObjectType<BookGetByIdQuery> {
        return {
            title: this.title
        };
    }
}
"
`;

exports[`Function transform: ValueObject 1`] = `
"import { UIntValue, R, OK, type JsonObjectType, type PlainParsableHObjectType, type PlainParseError, PlainParseHelper, InvalidTypePlainParseIssue, HObjectTypeMeta, PlainParseIssue, TooBigPlainParseIssue, TooSmallPlainParseIssue } from "@hexancore/common";
export class BookId extends UIntValue<BookId> {
    public static HOBJ_META = HObjectTypeMeta.domain("Book", "Book", "ValueObject", "BookId", BookId);
}
"
`;
